File: java/src/main/java/vsdk/toolkit/galerkin/ShadowCache.java
package vsdk.toolkit.galerkin;

import vsdk.toolkit.common.linealAlgebra.Numeric;
import vsdk.toolkit.common.linealAlgebra.Ray;
import vsdk.toolkit.material.RayHitFlag;
import vsdk.toolkit.skin.Patch;
import vsdk.toolkit.skin.RayHit;

public class ShadowCache {
    private static final int MAX_CACHE = 5;
    private Patch[] patchCache = new Patch[MAX_CACHE];
    private int numberOfCachedPatches;

    /**
Initialize/empty the shadow cache
*/
    public ShadowCache() {
        numberOfCachedPatches = 0;
        for ( int i = 0; i < MAX_CACHE; i++ ) {
            patchCache[i] = null;
        }
    }

    /**
Test ray against patches in the shadow cache. Returns nullptr if the ray hits
no patches in the shadow cache, or a pointer to the first hit patch otherwise
*/
    public RayHit cacheHit(Ray ray, float[] distance, RayHit hitStore) {
        for ( int i = 0; i < numberOfCachedPatches; i++ ) {
            if ( patchCache[i] == null ) {
                continue;
            }
            RayHit hit = patchCache[i].intersect(
                ray,
                Numeric.EPSILON_FLOAT * distance[0],
                distance,
                RayHitFlag.FRONT | RayHitFlag.ANY,
                hitStore);
            if ( hit != null ) {
                return hit;
            }
        }
        return null;
    }

    /**
Replace least recently added patch
*/
    public void addToShadowCache(Patch patch) {
        patchCache[numberOfCachedPatches % MAX_CACHE] = patch;
        if ( numberOfCachedPatches < MAX_CACHE ) {
            numberOfCachedPatches++;
        }
    }
}

File: typescript/src/main/vsdk/toolkit/galerkin/ShadowCache.ts
import { Numeric } from "../common/linealAlgebra/Numeric";
import { Ray } from "../common/linealAlgebra/Ray";
import { RayHitFlag } from "../skin/RayHitFlag";
import { Patch } from "../skin/Patch";
import { RayHit } from "../skin/RayHit";

export class ShadowCache {
  private static readonly MAX_CACHE = 5;
  private readonly patchCache: Array<Patch | null>;
  private numberOfCachedPatches: number;

  public constructor() {
    this.numberOfCachedPatches = 0;
    this.patchCache = new Array<Patch | null>(ShadowCache.MAX_CACHE).fill(null);
  }

  public cacheHit(ray: Ray, distance: number[], hitStore: RayHit): RayHit | null {
    for (let i = 0; i < this.numberOfCachedPatches; i++) {
      if (this.patchCache[i] === null) {
        continue;
      }
      const hit = (this.patchCache[i] as Patch).intersect(
        ray,
        Numeric.EPSILON_FLOAT * distance[0],
        distance,
        RayHitFlag.FRONT | RayHitFlag.ANY,
        hitStore,
      );
      if (hit !== null) {
        return hit;
      }
    }
    return null;
  }

  public addToShadowCache(patch: Patch): void {
    this.patchCache[this.numberOfCachedPatches % ShadowCache.MAX_CACHE] = patch;
    if (this.numberOfCachedPatches < ShadowCache.MAX_CACHE) {
      this.numberOfCachedPatches++;
    }
  }
}
